text/0000-dotdot-in-patterns.md

```diff
@@ -0,0 +1,124 @@
+- Feature Name: dotdot_in_patterns
+- Start Date: 2016-02-06
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Permit the `..` pattern fragment in more contexts.
+
+# Motivation
+[motivation]: #motivation
+
+The pattern fragment `..` can be used in some patterns to denote several elements in list contexts.
+However, it doesn't always compiles when used in such contexts.
+One can expect the ability to match tuple variants like `V(u8, u8, u8)` with patterns like
+`V(x, ..)` or `V(.., z)`, but the compiler rejects such patterns currently despite accepting
+very similar `V(..)`.
+
+This RFC is intended to "complete" the feature and make it work in all possible list contexts,
+making the language a bit more convenient and consistent.
+
+# Detailed design
+[design]: #detailed-design
+
+Let's list all the patterns currently existing in the language, that contain lists of subpatterns:
+
+```
+// Struct patterns.
+S { field1, field2, ..., fieldN }
+
+// Tuple struct patterns.
+S(field1, field2, ..., fieldN)
+
+// Tuple patterns.
+(field1, field2, ..., fieldN)
+
+// Slice patterns.
+[elem1, elem2, ..., elemN]
+```
+In all the patterns above, except for struct patterns, field/element positions are significant.
+
+Now list all the contexts that currently permit the `..` pattern fragment:
+```
+// Struct patterns, the last position.
+S { subpat1, subpat2, .. }
+
+// Tuple struct patterns, the last and the only position, no extra subpatterns allowed.
+S(..)
+
+// Slice patterns, the last position.
+[subpat1, subpat2, ..]
+// Slice patterns, the first position.
+[.., subpatN-1, subpatN]
+// Slice patterns, any other position.
+[subpat1, .., subpatN]
+// Slice patterns, any of the above with a subslice binding.
+// (The binding is not actually a binding, but one more pattern bound to the sublist, but this is
+// not important for our discussion.)
+[subpat1, binding.., subpatN]
+```
+Something is obviously missing, let's fill in the missing parts.
+
+```
+// Struct patterns, the last position.
+S { subpat1, subpat2, .. }
+// **NOT PROPOSED**: Struct patterns, any position.
+// Since named struct fields are not positional, there's essentially no sense in placing the `..`
+// anywhere except for one conventionally chosen position (the last one) or in sublist bindings,
+// so we don't propose extensions to struct patterns.
+S { subpat1, .., subpatN }
+S { subpat1, binding.., subpatN }
+
+// Tuple struct patterns, the last and the only position, no extra subpatterns allowed.
+S(..)
+// **NEW**: Tuple struct patterns, any position.
+S(subpat1, subpat2, ..)
+S(.., subpatN-1, subpatN)
+S(subpat1, .., subpatN)
+// **NEW**: Tuple struct patterns, any position with a sublist binding.
+// The binding has a tuple type.
+S(subpat1, binding.., subpatN)
```

Comment:
these are somewhat problematic. Afaik there's no guarantee that `(B, C)` has the same memory layout as the last two parts of `(A, B, C)`, so taking a slice into a tuple might not be so trivial.